### PR TITLE
[WIP] Improvements to reindexing

### PR DESCRIPTION
**Pull request by dantleech**
This PR aims to improve the search reindexing process.

- Reduce memory usage (note that the memory usage isn't significantly decreased in this PR, Zend Lucene takes alot).
- Allow resumption of indexing process.
- Better error notices.
- Show better progress: Realtime memory usage and current document.
- Now listener is fully unit tested.

## Related PRs

- Documentation: https://github.com/sulu-io/sulu-docs/pull/163
- MassiveSearch: https://github.com/massiveart/MassiveSearchBundle/pull/74

### CODE REVIEW

File: src/Sulu/Bundle/ContentBundle/Search/EventListener/ReindexListener.php [hunk 1]
```diff
@@ -94,59 +110,99 @@ public function onIndexRebuild(IndexRebuildEvent $event)
...
         );
 
-        $count = [];
 
+        $batchSize = 50;
+        $offset = $this->resumeManager->getCheckpoint(self::CHECKPOINT_NAME, 0);
+        $batchLimit = null;
+        $batchCount = 0;
+
+        do {
```
Comment (dantleech) [inline src/Sulu/Bundle/ContentBundle/Search/EventListener/ReindexListener.php]: If it is not too much overhead, we should show the number of documents to process..

File: src/Sulu/Bundle/ContentBundle/Search/EventListener/ReindexListener.php [hunk 2]
```diff
@@ -94,59 +110,99 @@ public function onIndexRebuild(IndexRebuildEvent $event)
...
-                } catch (\Exception $e) {
-                    $output->writeln(sprintf(
-                        '<error>Error indexing or de-indexing page (path: %s locale: %s)</error>: %s',
-                        $this->inspector->getPath($document),
-                        $locale,
-                        $e->getMessage()
-                    ));
                 }
+            } catch (\Exception $e) {
+                $error = '<error>Error indexing or de-indexing page</error>';
```
Comment (dantleech) [inline src/Sulu/Bundle/ContentBundle/Search/EventListener/ReindexListener.php]: Need to add way to show errors.

**Comment by dantleech** [comment]
Add docs indicating that reindexing should be done in `prod` and that reindexing with PHP7 is 10x faster.

File: src/Sulu/Bundle/ContentBundle/Search/EventListener/ReindexListener.php [hunk 3]
```diff
@@ -50,11 +55,17 @@ class ReindexListener
      */
     private $baseMetadataFactory;
 
+    /**
+     * @var RebuildResumeManager
```
Comment (dantleech) [inline src/Sulu/Bundle/ContentBundle/Search/EventListener/ReindexListener.php]: ResumeManager

File: src/Sulu/Bundle/ContentBundle/Search/EventListener/ReindexListener.php [hunk 4]
```diff
@@ -94,59 +111,114 @@ public function onIndexRebuild(IndexRebuildEvent $event)
             'SELECT * FROM [nt:unstructured] AS a WHERE ' . $condition
         );
 
-        $count = [];
+        $batchSize = 50;
+        $offset = $this->resumeManager->getCheckpoint(self::CHECKPOINT_NAME, 0);
+        $batchLimit = 4;
```
Comment (dantleech) [inline src/Sulu/Bundle/ContentBundle/Search/EventListener/ReindexListener.php]: set to `null` / remove

File: src/Sulu/Bundle/ContentBundle/Search/EventListener/ReindexListener.php [hunk 5]
```diff
@@ -94,59 +111,114 @@ public function onIndexRebuild(IndexRebuildEvent $event)
...
+        $batchCount = 0;
+
+        // index in batches: indexing in batches means:
+        //
+        //   1. Jackalope does not need to hydrate all of the PHPCR nodes from the query.
+        //   2. We can clear the document manager registry periodically.
```
Comment (dantleech) [inline src/Sulu/Bundle/ContentBundle/Search/EventListener/ReindexListener.php]: Disclaimer: neither of these points drastically reduces memory usage.

File: src/Sulu/Bundle/ContentBundle/Search/EventListener/ReindexListener.php [hunk 6]
```diff
@@ -94,59 +111,114 @@ public function onIndexRebuild(IndexRebuildEvent $event)
...
-        $output->writeln('');
+            $output->write(' Mem: ' . number_format(memory_get_usage()) . 'b');
 
-        foreach ($count as $className => $count) {
-            if ($count == 0) {
-                continue;
+            if ($document instanceof TitleBehavior) {
+                $output->write(' Title: ' . $document->getTitle(). "\x1B[0J");
+            } else {
+                $output->write(' OID: ' . spl_object_hash($document->getTitle()). "\x1B[0J");
```
Comment (dantleech) [inline src/Sulu/Bundle/ContentBundle/Search/EventListener/ReindexListener.php]: Test or remove.

File: src/Sulu/Bundle/ContentBundle/Search/EventListener/ReindexListener.php [hunk 7]
```diff
@@ -94,59 +111,118 @@ public function onIndexRebuild(IndexRebuildEvent $event)
...
+                $documents->next();
                 $progress->advance();
                 continue;
             }
 
-            $locales = $this->inspector->getLocales($document);
+            try {
+                $locales = $this->inspector->getLocales($document);
+            } catch (\Exception $e) {
+                $this->logError($output, 'Error indexing page', $e);
```
Comment (dantleech) [inline src/Sulu/Bundle/ContentBundle/Search/EventListener/ReindexListener.php]: .. and continue

File: src/Sulu/Bundle/ContentBundle/Search/EventListener/ReindexListener.php [hunk 8]
```diff
@@ -94,59 +111,121 @@ public function onIndexRebuild(IndexRebuildEvent $event)
...
-        $output->writeln('');
+            $output->write(' Mem: ' . number_format(memory_get_usage()) . 'b');
 
-        foreach ($count as $className => $count) {
-            if ($count == 0) {
-                continue;
+            if ($document instanceof TitleBehavior) {
+                $output->write(' Title: ' . $document->getTitle() . "\x1B[0J");
+            } else {
+                $output->write(' OID: ' . spl_object_hash($document) . "\x1B[0J");
```
Comment (dantleech) [inline src/Sulu/Bundle/ContentBundle/Search/EventListener/ReindexListener.php]: Can use `DocumentHelper::getDebugTitle` here..

**Comment by dantleech** [comment]
Ready. @chirimoya @danrot @wachterjohannes

File: src/Sulu/Bundle/ContentBundle/Search/EventListener/ReindexListener.php [hunk 9]
```diff
@@ -94,59 +111,121 @@ public function onIndexRebuild(IndexRebuildEvent $event)
...
+
+            // this reduces overall memory usage by 1% in one experiment, so is not essential
+            $this->documentManager->clear();
+            $offset += $batchSize;
+            ++$batchCount;
+
+            if (null !== $batchLimit && $batchCount > $batchLimit) {
+                break;
+            }
+            $this->resumeManager->setCheckpoint(self::CHECKPOINT_NAME, $offset);
```
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/Search/EventListener/ReindexListener.php]: As stated previously I am not really happy with this solution... Let's see what @chirimoya has to add.

**Comment by dantleech** [comment]
As discussed, in order to allow resumption of indexing jobs on the ORM I will need to introduce a query builder which can be modified by clients from their Entity repositories.

File: src/Sulu/Bundle/ContentBundle/Search/EventListener/ReindexListener.php (hunk 9 above)
Comment (webmozart) [inline src/Sulu/Bundle/ContentBundle/Search/EventListener/ReindexListener.php]: Why not?

Comment (dantleech) [inline src/Sulu/Bundle/ContentBundle/Search/EventListener/ReindexListener.php]: See #2018 - basically the resume manager was distributed, as was lots of logic. As of *this* PR we simply issued a `reindex` event and let the subscribers (there are two: Doctrine ORM and Sulu Structure) do whatever they need in order to reindex content.

This obviously limited how much control we had over the process, and led to distributing more responsiblities.

The #2018 PR (made against this branch) changes the situation, so the equivilent of this class simply provides the data to be reindexed.

Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/Search/EventListener/ReindexListener.php]: @webmozart The crucial part for me was that this class had to use the `ResumeManager` on its own. So if the user who extends the MassiveSearchBundle this way forgets about that, the resuming doesn't work as expected.

**Comment by dantleech** [comment]
Rebased.

File: CHANGELOG.md [hunk 10]
```diff
@@ -5,6 +5,7 @@ CHANGELOG for Sulu
     * ENHANCEMENT #2014 [Content]             Allow `-` in webspace name.
     * ENHANCEMENT #1999 [SnippetBundle]       Snippet controller now uses DocumentManager and Serializer.
     * ENHANCEMENT #2008 [ContactBundle]       Fixed sorting in contact selection content type
+    * ENHANCEMENT #1981 [ConetntBundle]       Better search reindexing for structure content.
```
Comment (danrot) [inline CHANGELOG.md]: `ContentBundle`

File: src/Sulu/Bundle/ContentBundle/Search/ReIndex/StructureProvider.php [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+    }
+
+    public function getCount($classFqn)
+    {
+    }
+
+    public function getClassFqns()
+    {
+    }
+}
```
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/Search/ReIndex/StructureProvider.php]: This file should probably be gone :-)

Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/Search/ReIndex/StructureProvider.php]: The existence of this folder also seems to produce a case mismatch.

**Pull request merged**